### PR TITLE
move region peer_cache to the peer cache itself

### LINKED ISSUE #1843: consider using vector instead of hashmap in raft Peer
Now we use many HashMap in raft Peer, like peer cache,  peer heartbeats and node prs. In most of the case, the hash map is very small (we can guarantee that the size is about the max replica number 3 or a little larger 4). So we can use vector instead, for peer cache, we can move the region peer cache to the peer cache itself, no need to maintain a total peer cache in store. 

```
#[bench]
fn bench_vec_find_1(b: &mut Bencher) {
    let v = vec![(1, 1), (2, 2), (3, 3), (4, 4), (5, 5)];
    b.iter(||v.iter().find(|&x| x.0 == 1).unwrap())
}

#[bench]
fn bench_vec_find_5(b: &mut Bencher) {
    let v = vec![(1, 1), (2, 2), (3, 3), (4, 4), (5, 5)];
    b.iter(||v.iter().find(|&x| x.0 == 5).unwrap())
}

#[bench]
fn bench_ordermap_find_1(b :&mut Bencher) {
    let m = ordermap!{1 => 1, 2 => 2, 3 => 3, 4 => 4, 5 => 5};
    b.iter(||m[&1])
}

#[bench]
fn bench_ordermap_find_5(b :&mut Bencher) {
    let m = ordermap!{1 => 1, 2 => 2, 3 => 3, 4 => 4, 5 => 5};
    b.iter(||m[&5])
}

#[bench]
fn bench_hashmap_find_1(b :&mut Bencher) {
    let mut m = HashMap::new();
    m.insert(1, 1);
    m.insert(2, 2);
    m.insert(3, 3);
    m.insert(4, 4);
    m.insert(5, 5);    
    b.iter(||m[&1])
}

#[bench]
fn bench_hashmap_find_5(b :&mut Bencher) {
    let mut m = HashMap::new();
    m.insert(1, 1);
    m.insert(2, 2);
    m.insert(3, 3);
    m.insert(4, 4);
    m.insert(5, 5);    
    b.iter(||m[&5])
}
```
```
test tests::bench_hashmap_find_1  ... bench:          22 ns/iter (+/- 7)
test tests::bench_hashmap_find_5  ... bench:          19 ns/iter (+/- 13)
test tests::bench_ordermap_find_1 ... bench:          25 ns/iter (+/- 11)
test tests::bench_ordermap_find_5 ... bench:          30 ns/iter (+/- 17)
test tests::bench_vec_find_1      ... bench:           2 ns/iter (+/- 0)
test tests::bench_vec_find_5      ... bench:           7 ns/iter (+/- 6)
```

As we can see, for small data search, vector is very faster than hash map.

### PR DESCRIPTION
**Pull request by javaforfun**
Fixes #1843.

### CODE REVIEW

**Comment by siddontang** [comment]
Any benchmark?

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -1087,10 +1073,6 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         };
 
         self.region_peers.get_mut(&region_id).unwrap().mut_store().region = origin_region.clone();
-        for peer in new_region.get_peers() {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: If this would broken the logic?

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
I suggest moving peer_cache to another PR.

**Comment by javaforfun** [comment]
OK,  I will submit 2 new PR,  and close this PR.

**Comment by javaforfun** [comment]
please merge PR "replace hashmap raft.prs with flatmap" first.

**Comment by ngaut** [comment]
Please rebase master.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -192,14 +192,14 @@ pub struct PeerStat {
 pub struct Peer {
     engine: Arc<DB>,
     cfg: Rc<Config>,
-    peer_cache: Rc<RefCell<HashMap<u64, metapb::Peer>>>,
+    peer_cache: RefCell<FlatMap<u64, metapb::Peer>>,
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: If it's not going to be shared, then `RefCell` is unnecessary.

Comment (javaforfun) [inline src/raftstore/store/peer.rs]: use RefCell for interior mutability, as get_peer_from_cache may insert new elem( ../blob/master/src/raftstore/store/store.rs#L1305)

Comment (BusyJay) [inline src/raftstore/store/peer.rs]: I think it can be borrowed mutably.

**Comment by BusyJay** [comment]
LGTM

File: src/raftstore/store/peer.rs (hunk 2 above)
Comment (siddontang) [inline src/raftstore/store/peer.rs]: can we remove RefCell, @BusyJay ?

Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Yes. But may not in this pr.

Comment (siddontang) [inline src/raftstore/store/peer.rs]: @javaforfun please update it later.

**Pull request merged**